File: src/test/java/org/mapdb/MapListenerTest.java
package org.mapdb;

import org.junit.Test;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@SuppressWarnings({"rawtypes","unchecked"})
public class MapListenerTest {

        @Test public void hashMap(){
            tt(DBMaker.memoryDB().transactionDisable().cacheHashTableEnable().make().hashMap("test"), false);
        }

        @Test public void treeMap(){
            tt(DBMaker.memoryDB().transactionDisable().cacheHashTableEnable().make().treeMap("test"), false);
        }
        
        @Test public void hashMapAfter(){
            tt(DBMaker.memoryDB().transactionDisable().cacheHashTableEnable().make().hashMap("test"), true);
        }

        @Test public void treeMapAfter(){
            tt(DBMaker.memoryDB().transactionDisable().cacheHashTableEnable().make().treeMap("test"), true);
        }


        void tt(Bind.MapWithModificationListener m, boolean after){
            final AtomicReference key = new AtomicReference(null);
            final AtomicReference newVal = new AtomicReference(null);
            final AtomicReference oldVal = new AtomicReference(null);
            final AtomicInteger counter = new AtomicInteger(0);

            Bind.MapListener listener = new Bind.MapListener(){
                @Override public void update(Object key2, Object oldVal2, Object newVal2) {
                    counter.incrementAndGet();
                    key.set(key2);
                    oldVal.set(oldVal2);
                    newVal.set(newVal2);
                }
            };

            if (after){
                m.modificationListenerAfterAdd(listener);
            }else{
                m.modificationListenerAdd(listener);
            }
            

            //check CRUD
            m.put("aa","bb");
            assertTrue(key.get()=="aa" && newVal.get()=="bb" && oldVal.get()==null && counter.get()==1);

            m.put("aa","cc");
            assertTrue(key.get()=="aa" && newVal.get()=="cc" && oldVal.get()=="bb" && counter.get()==2);

            m.remove("aa");
            assertTrue(key.get()=="aa" && newVal.get()==null && oldVal.get()=="cc" && counter.get()==3);

            if (!after){
                //check clear()
                m.put("aa","bb");
                assertTrue(key.get()=="aa" && newVal.get()=="bb" && oldVal.get()==null && counter.get()==4);
                m.clear();
                assertTrue(key.get()=="aa" && newVal.get()==null && oldVal.get()=="bb" && counter.get()==5);
            }

            //check it was unregistered
            counter.set(0);
            if (after){
                m.modificationListenerAfterRemove(listener);
            }else{
                m.modificationListenerRemove(listener);
            }
            m.put("aa","bb");
            assertEquals(0, counter.get());
        }

}

File: src/test/java/org/mapdb/MapListenerTest.kt
package org.mapdb

import kotlin.test.assertEquals
import org.junit.*
import kotlin.test.assertTrue


class MapListenerTest{

    Test fun hashMap(){
        tt(DBMaker.newMemoryDB().journalDisable().make().getHashMap("test"))
    }

    Test fun treeMap(){
        tt(DBMaker.newMemoryDB().journalDisable().make().getTreeMap("test"))
    }


    fun tt(m:Bind.MapWithModificationListener<Any?,Any?>){
        var key:Any? = null
        var newVal:Any? = null
        var oldVal:Any? = null
        var counter = 0;

        val listener = mapListener{(key2:Any?,oldVal2:Any?,newVal2:Any?) ->
           counter++;
           key = key2;
           oldVal = oldVal2;
           newVal = newVal2;
        }

        m.addModificationListener(listener);

        //check CRUD
        m.put("aa","bb")
        assertTrue(key=="aa" && newVal=="bb" && oldVal==null && counter==1);

        m.put("aa","cc")
        assertTrue(key=="aa" && newVal=="cc" && oldVal=="bb" && counter==2);

        m.remove("aa")
        assertTrue(key=="aa" && newVal==null && oldVal=="cc" && counter==3);

        //check clear()
        m.put("aa","bb")
        assertTrue(key=="aa" && newVal=="bb" && oldVal==null && counter==4);
        m.clear()
        assertTrue(key=="aa" && newVal==null && oldVal=="bb" && counter==5);


        //check it was unregistered
        counter = 0;
        m.removeModificationListener(listener);
        m.put("aa","bb");
        assertEquals(0, counter);
    }
}
